### DISCUSSION TITLE
convergence issue when SSCHA matrix has a phonon mode close to zero

### DISCUSSION THREAD

**Discussion by kw600** [SSCHAcode/python-sscha#187] [Q&A] [answered]
Dear SSCHA developers,

The SSCHA matrix is positive definite. There is chance one soft phonon mode is very close to zero (small positive value) at some specific temperature, which makes the SSCHA matrix has a phonon frequency close to zero. This results in the minimization of Force constants matrix flctuating aroung several hundred and never going down. Is there any way to get over this?

Thank you for your help.
Kang

**Comment by mesonepigreco** [comment, marked as answer]
Hi Kang,
Yes, this could be the case, especially if the temperature is low (below 100 K).
In these cases, it is often good to try minimizing the free energy from a higher temperature (100 or 200 K above the target) and then lowering the temperature progressively (use the higher temperature as a starting point for a lower temperature calculation).
A higher temperature may help as it populates the small modes more, generating a larger average displacement of ions along the low-frequency vibrations that helps the algorithm sample the anharmonicity better. 

The fluctuations of the FC matrix could also be due to the insufficient number of configurations, especially in systems with few symmetries (low symmetry group, many atoms in the primitive cell, etc.). In this case, increasing them can solve the issue, and fitting a force field on the partial data obtained from ab initio could be a solution (see, for example, the tutorial https://sscha.eu/Tutorials/tutorial_06_the_SSCHA_with_MLP/)

Then, it may depend on the specific case, so if you think the issue cannot be solved in these ways, feel free to write more details so we can help you further.

All the best,
Lorenzo

**Reply by kw600** [reply]
Dear Lorenzo,

Many thanks for your reply.

By increasing the number of configurations and reducing temperature gradually, the force constants matrix finally becomes converged. Your suggestions are very helpful.

I just have another question about the code. I am now using the automatic mode to run SSCHA.
relax = sscha.Relax.SSCHA(minimizer, espresso_config,
        N_configs=config.n_config,
        max_pop=20,
        save_ensemble=True,
        cluster=cluster_config)
relax.vc_relax(target_press = 0,static_bulk_modulus = 300,ensemble_loc = "ensembles")

The code can generate populations and submit them to the cluster and get the results back.
I am just wondering, after the free energy is converged, how can I automatically ask the code to use the cluster to calculate the free energy hessian? (I ususally mannually do this after I get the final SSCHA matrix.)
I see in the tutorial it is ensemble.get_energy_forces(ff_calculator, compute_stress = False), which are not calculated locally.

Best wishes,
Kang

**Reply by mesonepigreco** [reply]
Hi, if you use instead of ``ensemble.get_energy_forces`` the function ``ensemble.compute_ensemble`` you can also pass the cluster object to it to send the calculation on a remote cluster.
See https://sscha.eu/python-sscha.pdf, section 6.1, for a pdf small description of this function, or directly the docstring for a complete list of arguments.
```
def compute_ensemble(self, calculator, compute_stress = True, stress_numerical = False,
                         cluster = None, verbose = True, timer=None):
        """
        GET ENERGY AND FORCES
        =====================

        This is the generic function to compute forces and stresses.
        It can be used both with clusters, and with simple ase calculators

        Paramters
        ---------
            calculator:
                The ase calculator
            compute_stress: bool
                If true compute the stress
            stress_numerical : bool
                Compute the stress tensor with finite difference,
                this is not possible with clusters
            cluster: Cluster, optional
                The cluster in which to send the calculation.
                If None the calculation is performed on the same computer of
                the sscha code.
        """
```

In particular, you need to pass the cluster (sscha.Cluster.Cluster) object and the calculator (the same you used to run the automatic Relax, to the compute_ensemble function to automatically submit the calculation on the remote cluster. 

So you need to add an extra part to your script in which the code generates one last (larger) population after the minimization, it does a one-shot minimization, and then evaluates the free energy Hessian, and saves the results (essentially what the tutorial does).

**Reply by kw600** [reply]
Dear Lorenzo,

Thank you for your help. It works now. 

I may find a possible bug. After using the compute_ensemble with cluster. I can see the ensemble.force correctly store all the calculated results and therefore the ensemble.save_bin can save these results into .npy format. However, it seems the flag of ensemble.force_calculated is not set to be True so that the function ensemble.save does not save all the forces.

Best wishes,
Kang
